Refactored ScheduledJobUIViewSet (approval_queue view incomplete – he…

Pavan-183 opened the pull request:
…lp needed)

<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes #<ISSUE NUMBER GOES HERE>
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
-->
# Screenshots
<!--
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

Pavan-183 commented:
![image](https://github.com/user-attachments/assets/a3fa7711-3384-4adb-a233-5d959e1c0fda)

Hi team, I'm facing this issue while accessing the **approval_queue**, I'm not entirely sure if the approach I've taken is correct. Could someone please review the logic and suggest improvements?

glennmatthews commented on nautobot/extras/templates/extras/job_approval_request.html:
```diff
@@ -122,7 +122,7 @@ <h1>{{ object.name }}</h1>
                                 <i class="mdi mdi-close"></i> Deny
                             </button>
-                            <a href="{% url 'extras:scheduledjob_approval_queue_list' %}" class="btn btn-default">
+                            <a href="extras/jobs/approval-queue" class="btn btn-default">
```
This change is a bad code smell. We should still be able to use `{% url ... %}` here.

glennmatthews commented:
Hard to tell what the issue is in your screenshot without the traceback for context.

Pavan-183 commented:
> Hard to tell what the issue is in your screenshot without the traceback for context.

Hi @glennmatthews , I have added the following logic for the `approval_queue` custom view inside `ScheduledJobUIViewSet`:
However, I’m facing a rendering issue ('str' object has no attribute '_meta'), which seems related to the context or how the template is using the view. I’d appreciate your thoughts on whether this usage aligns with the Nautobot UI framework, or if there's a better way to structure this.

    @action(detail=False, url_path="approval-queue")
    def approval_queue(self, request, *args, **kwargs):
        queryset = ScheduledJob.objects.needs_approved()
        table = tables.ScheduledJobApprovalQueueTable(queryset)
        context = {
            "view": self,
            "queryset": queryset,
            "table": table,
            "filter_form": self.filterset_form_class(request.GET or None),
        }
        return render(request, "extras/scheduled_jobs_approval_queue_list.html", context)
        
I have added this logic

itdependsnetworks commented on nautobot/extras/views.py:
```diff
@@ -1983,52 +1983,40 @@ def destroy(self, request, *args, **kwargs):
+    @action(detail=False, url_path="approval-queue")
+    def approval_queue(self, request, *args, **kwargs):
+        queryset = ScheduledJob.objects.needs_approved()
+        table = tables.ScheduledJobApprovalQueueTable(queryset)
```
Just taking a guess, but perhaps remove the callable to the function, as logic down the road may be doing that.

HanlinMiao commented:
The more I look at this, the more I think this might be the correct approach https://github.com/nautobot/nautobot/pull/7255/commits/13e778f34a72f75c3c176359b30311bae852701a

Pavan-183 commented:
> The more I look at this, the more I think this might be the correct approach [13e778f](https://github.com/nautobot/nautobot/commit/13e778f34a72f75c3c176359b30311bae852701a)

Okay @HanlinMiao .

itdependsnetworks commented on nautobot/extras/urls.py:
```diff
@@ -319,17 +320,9 @@
     path(
         "jobs/scheduled-jobs/approval-queue/",
-        views.ScheduledJobApprovalQueueListView.as_view(),
+        views.ScheduledJobApprovalQueueListView.as_view({"get": "list"}),
```
I don't fully get this, but will defer to core to understand a bit better.

glennmatthews replied:
Not sure I understand the need for this either :-)

HanlinMiao replied:
So if you use a standalone `ObjectListViewMixin` and not register it with `router.register()`. 
The error below would appear, as we are missing [this](https://github.com/nautobot/nautobot/blob/5f30141bcf28cd43b9a95723b499a7f8394e54c1/nautobot/core/views/routers.py#L9-L16) in the `NautobotUIViewSetRouter`. Similar patterns can be found [here](https://github.com/nautobot/nautobot/blob/5f30141bcf28cd43b9a95723b499a7f8394e54c1/examples/example_app/example_app/urls.py#L40-L44) and [here](https://github.com/nautobot/nautobot/blob/5f30141bcf28cd43b9a95723b499a7f8394e54c1/nautobot/circuits/urls.py#L23-L27).

```
celery_beat-1    |     views.ScheduledJobApprovalQueueListView.as_view(),
celery_beat-1    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_beat-1    |   File "/usr/local/lib/python3.11/site-packages/rest_framework/viewsets.py", line 83, in as_view
celery_beat-1    |     raise TypeError("The `actions` argument must be provided when "
celery_beat-1    | TypeError: The `actions` argument must be provided when calling `.as_view()` on a ViewSet. For example `.as_view({'get': 'list'})`
```

itdependsnetworks commented on nautobot/extras/views.py:
```diff
@@ -1984,42 +1984,32 @@ def destroy(self, request, *args, **kwargs):
 
 
-class ScheduledJobListView(generic.ObjectListView):
+class ScheduledJobUIViewSet(
```
Can you mark up which is missing as a comment and why it's not there. Makes it easier to understand down the road.

glennmatthews replied:
At a glance I *think* the only missing one is `ObjectCreateViewMixin`, which makes sense, but agreed, comments would be helpful.

glennmatthews commented on nautobot/extras/views.py:
```diff
@@ -2028,8 +2018,17 @@ def get_extra_context(self, request, instance):
+    def get_template_name(self):
+        if self.action == "list":
+            return "extras/scheduled_jobs_approval_queue_list.html"
+        return super().get_template_name()
```
When (if ever) will the action *not* be `list`?

itdependsnetworks replied:
If that is the case I think it would make sense to auto raise an error, something like 

```python
if self.action != "list":
    raise Something("some descriptive message")
```

glennmatthews commented on nautobot/extras/views.py:
```diff
@@ -1984,42 +1984,32 @@ def destroy(self, request, *args, **kwargs):
+class ScheduledJobUIViewSet(
+    mixins.ObjectDetailViewMixin,
+    mixins.ObjectListViewMixin,
+    mixins.ObjectEditViewMixin,
+    mixins.ObjectDestroyViewMixin,
+    mixins.ObjectBulkDestroyViewMixin,
+    mixins.ObjectBulkCreateViewMixin,
+    mixins.ObjectChangeLogViewMixin,
+    mixins.ObjectNotesViewMixin,
```
Actually, I don't think `Edit`, `BulkCreate` apply to ScheduledJobs at present, and as `ScheduledJob` is a `BaseModel` only, I don't think `ChangeLog` or `Notes` apply either?

HanlinMiao commented:
Close due to potential conflict with remaining work with `ApprovalWorkflow` feature related to `ScheduledJob` in `next`
